### ISSUE TITLE
On-the-Fly Project and Activity Creation with Random Colors

### ISSUE THREAD

**Issue by SamirSaidani** [kimai/kimai#4439]
### Describe the problem

It's cumbersome to create a project and/or an activity as a volunteer for a non-profit org, since you leave all the values to default (zero and no invoice since you are volunteer).

### Describe the solution you'd like

It would be awesome if, while creating a new entry, I could also create a Project and/or Activity on the fly directly from the entry record of the timesheet page. Additionally, assigning random colors to these new projects and activities would add a nice touch. It would probably be useful too for a for-profit org, since it's an easy way to go fast and take care of the invoice details later.

**Comment by SamirSaidani** [comment]
I'm using Kimai for a non-profit organization, assessing its viability for daily use. As I delve into Kimai, I find the software's functionality quite impressive, yet I see room for improvement in terms of UX. In general, I'm looking for a smoother UX experience rather than having the interface standing on my way.

In the "My Times" screen, I'd like volunteers to have the ability to create their own projects and activities on-the-fly, as shown here:

![image](https://github.com/kimai/kimai/assets/842380/94110d2e-4301-4e38-b2cf-c58bcf17b0d2)

![image](https://github.com/kimai/kimai/assets/842380/15ee1af4-9b15-49c6-ade1-dd9219ef93be)

Additionally, for 'No results found for 'My awesome project,'' we could add a second line saying 'Create this project,' which would create the project with the entered name and leave all other values as default. Furthermore, a third line could be added saying 'Create and edit this project,' which would create the project with the entered name and open a second window over the first for editing the project.

The same functionality could be applied to activities if the admin grants permission to the user. A similar feature exists in Odoo in various parts of the GUI: if an item is not found, it can be created, or created and edited on-the-fly.

**Comment by kevinpapst** [comment]
While I like the idea, it is not as easy as it looks. 
Opening a second modal is not really feasible.
Creating it on the fly (and reloading the dropdown) might work, but Projects have the mandatory connection to customer.
In your case, you don't see that dropdown, because there is only one.

Customers have some mandatory fields as well. We have default settings for them, but still...

And then activities: two options:
- create activity for project XYZ
- create global activity

**Comment by SamirSaidani** [comment]
> While I like the idea, it is not as easy as it looks. Opening a second modal is not really feasible. Creating it on the fly (and reloading the dropdown) might work, but Projects have the mandatory connection to customer. In your case, you don't see that dropdown, because there is only one.



I see. To provide some context, we utilize Kimai to manage our company, which operates as a freelancer cooperative platform. It is owned and managed by freelancers, functioning both as a platform for freelancers and as a client for freelancers. As a common scenario, freelancers may employ Kimai to track projects and activities that support their own company.

I suggest the following:
- allow users to designate a default client, which would represent their own company,
- enable users to create a special client with a 'Default' checkbox. Include an explanation such as: 'You can utilize this option if you have a major client or if you wish to track time for your own company using Kimai.'

> Customers have some mandatory fields as well. We have default settings for them, but still...

Adding a 'Default Customer' resolves the issue, and the 'On-the-fly' feature only applies to the default customer.

> And then activities: two options:
> 
>     * create activity for project XYZ
> 
>     * create global activity

If the user creates 'My Activity' without filling in 'My Project', it will be automatically considered a global activity. 
if the user creates 'My Project' and then 'My Activity', 'My Activity' will be automatically linked to 'My Project'.

**Comment by R-Hammer** [comment]
Looking for that feature too!
Here its somehow possible: https://www.kimai.org/store/alexandreptj-kemai-desktop-client.html

**Comment by azerbinati** [comment]
Hi, I tried using Kimai to manage my worked hours. I regularly use different tracking systems like Redmine, Jira, and GitHub, which my clients use to indicate tasks to be done. I thought of using Kimai to agilely record the time worked for various clients, but I found out that both project creation, activity, and even tags cannot be created on the fly. I work on hundreds of different tickets, and every day I come across dozens of new ones. The only place where I can insert a reference is in the description, but this doesn't allow me to summarize the hours worked as I would like. I'm considering different systems, but if Kimai maintains these constraints, it's not suitable for my work scenario. I understand that creating a project requires filling in mandatory data, but activities and tags shouldn't. In any case, if you decide to develop this feature, it would be extremely important for it to simply allow entering plain text, which is currently 'not found.' Adding a modal screen to input details would still be a hindrance.

**Comment by kevinpapst** [comment]
@azerbinati Tags can be created on the fly. 
For the other points: test the custom-fields feature on the demo installation.
From reading your posts I would say you want to create a field "external reference" where you add your ticket ID.

> I work on hundreds of different tickets

In what time range? Don't need an answer, just wanted to point out that exaggeration doesn't help any discussion.
And activities were never meant to be highly dynamic.

**Comment by azerbinati** [comment]
> @azerbinati Tags can be created on the fly. 

You are right @kevinpapst! I was fooled by the "no results found" when you start typing. Didn't notice that I could create one after entering at least 3 characters. Anyway I really need activities because it's not just a "tag" but a description line.
My issues are like this "Feature #28802: Implementazioni piani di fatturazione da OdC"
Having tag "28802" could work.. but it's not human friendly.

>For the other points: test the custom-fields feature on the demo installation. From reading your posts I would say you want to >create a field "external reference" where you add your ticket ID.

I'll give it a try! Thank you.

> 
> > I work on hundreds of different tickets
> 
> In what time range? Don't need an answer, just wanted to point out that exaggeration doesn't help any discussion. And activities were never meant to be highly dynamic.

Perhaps it might aid our discussion if I elaborate on my background a bit more. I'm not just a task executor; I'm a team leader. My role involves initiating Issues that others will tackle. Understanding this could clarify why the volume of tasks, which might seem overwhelming to you, doesn't feel as burdensome with this perspective in mind.

In addition to this, my daily routine includes analyzing, describing, and assigning tasks (i.e., explaining the issue to the assignee), providing mentorship and support, and ultimately conducting code reviews of the work completed by my team. Typically, I dedicate about 5/6 hours a day to these coordination efforts across numerous issues, helping others to complete their tasks. The remaining time, which is limited, is spent on development tasks, much like the rest of the team members.

Previously, I would log all this time as an "internal activity" in our company's time-tracking system. However, this approach is flawed because approximately 50% of that time is actually billable to our clients. Hence, I need to allocate my time to the specific issues if I anticipate we can be billing for that time. Until now, I've balanced this by logging some activities on specific issues and others as "generic internal activity". Nonetheless, this process is cumbersome, and I'd much prefer to log my time directly on the relevant issues, even if it's for a brief 2-minute response on Teams, as that's reflective of most of my day.

With this in mind, I hope you better understand why it's important to me spend no time in activity creation.

To be honest, I've also tried using Clockify, which has been a good fit for me, but I require a solution that can be hosted on-premise and possibly be modified (contributed) by us.

What i really like in Clockify is that the work flow is inverted: you can start the time just clicking "start" and fill in description, project, activity, later whenever you like. In addition, you can simply click start on an already present record to start a new identical time record without any additional input and change if needed any detail later. (just think you have a few team mates that call you every 20/30 minutes asking help for the task they are on and you understand why it's a good value to be able to re-start a timer later ) 

Sorry if the last part drove the discussion a little bit off-topic, but the core of the original request is just to let "starting the chrono" have less requirements and be more agile.

Wish you best and sorry for the long post.

**Comment by kevinpapst** [comment]
Thanks for the detailed explanation.

I agree that the Clockify approach can be nice in certain situations, but I definitely know that many customers would protest if project and activity became optional. 

Right now the entire core assumes that project and activity are set. 
I'd love to have the option to make it behavior customizable, but this would be the first point that needs to be adapted. 

> In addition, you can simply click start on an already present record to start a new identical time record without any additional input

That works from the context menu in the "My times" view.

**Comment by robot-rubik** [comment]
had the same problem. built https://invoicechaser-app.surge.sh after chasing a client for 3 months on a $2k invoice. just automated the follow-up sequence so i don't have to decide how firm to be each time.
